File: DynamicArray/DynamicArray.java
package DynamicArray;

class DArray <T extends Comparable<T>> {
  private static final int INITIAL_SIZE = 16;
  private static final int RESIZE_FACTOR = 2;

  private int size, capacity;
  private T[] data;

  public DArray() {
    this(INITIAL_SIZE);
  }
  
  public DArray(int capacity) {
    size = 0;
    this.capacity = capacity > 0 ? capacity : INITIAL_SIZE;
    data = (T[])new Comparable[capacity];
  }

  public int getSize() {
    return size;
  }

  public int getCapacity() {
    return capacity;
  }

  public void add(T value) {
    // Resize
    if(size + 1 == capacity) {
      capacity *= RESIZE_FACTOR;
      T[] newData = (T[])new Comparable[capacity];

      for(int i = 0; i < size; ++i) 
        newData[i] = data[i];

      data = newData;
      data[size++] = value;

      newData = null;
    }
    else {
      data[size++] = value;
    }
  }

  public T remove(int index) {
    if(index < 0 || index >= size) throw new IndexOutOfBoundsException();

    T value = data[index];

    for(int i = index + 1; i < size; ++i) {
      data[i - 1] = data[i];
    }

    data[--size] = null;

    return value;
  }

  public T get(int index) {
    if(index < 0 || index >= size) throw new IndexOutOfBoundsException();
    
    return data[index];
  }

  public int indexOf(T value) {
    for(int i = 0; i < size; ++i) {
      if(data[i].equals(value))
        return i;
    }

    return -1;
  }

  public boolean contains(T value) {
    return indexOf(value) != -1;
  }

  @Override
  public String toString() {
    StringBuilder builder = new StringBuilder();
    builder.append("[");
    for(int i = 0; i < size; ++i) 
      builder.append(data[i]).append(" ");
    builder.delete(builder.length() - 1, builder.length()).append("]");

    return builder.toString();
  }
}

public class DynamicArray {
  public static void main(String[] args) {
    DArray<Integer> arr = new DArray<Integer>(2);
    arr.add(5);
    arr.add(3);
    arr.add(7);
    arr.add(512);
    arr.add(55);
    arr.add(124);

    System.out.println(arr);

    System.out.println(arr.indexOf(512));
    arr.remove(arr.indexOf(512));
    System.out.println(arr);
    System.out.println(arr.contains(124));
    System.out.println(arr.getCapacity());
    System.out.println(arr.getSize());
    System.out.println(arr.get(2));
  }
}

File: DynamicArray/DynamicArray.cpp
#include <iostream>

template<typename T>
class DArray {
  private:
    static const int INITIAL_SIZE = 16;
    static const int RESIZE_FACTOR = 2;

    int size, capacity;
    T* data;

  public:
    DArray() : DArray(INITIAL_SIZE) { }

    DArray(int capacity) : size{0} {
      this->capacity = capacity > 0 ? capacity : INITIAL_SIZE;
      data = new T[capacity];
    }

    DArray(const DArray<T>& other) {
      capacity = other.capacity;
      size = other.size;
      data = new T[capacity];

      for(int i = 0; i < size; ++i) 
        data[i] = other.data[i];
    }

    DArray(const DArray<T>&& other) {      
      capacity = other.capacity;
      size = other.size;
      data = other.data;

      other.data = nullptr;
    }

    DArray<T>& operator=(const DArray<T>& other) {
      if(*this == other) return *this;

      capacity = other.capacity;
      size = other.size;

      delete[] data;
      data = new T[capacity];

      for(int i = 0; i < size; ++i) 
        data[i] = other.data[i];

      return *this;
    }

    DArray<T>& operator=(const DArray<T>&& other) {
      if(*this == other) return *this;

      capacity = other.capacity;
      size = other.size;
      data = other.data;

      other.data = nullptr;
    }

    ~DArray() {
      delete[] data;
    }

    int getSize() const {
      return size;
    }

    int getCapacity() const {
      return capacity;
    }

    void add(T value) {
      // Resize
      if(size + 1 == capacity) {
        capacity *= RESIZE_FACTOR;
        T* newData = new T[capacity];

        for(int i = 0; i < size; ++i) 
          newData[i] = data[i];

        delete[] data;

        data = newData;
        data[size++] = value;
      }
      else {
        data[size++] = value;
      }
    }

    T remove(int index) {
      if(index < 0 || index >= size) throw std::runtime_error("Index out of bounds");

      T value = data[index];

      for(int i = index + 1; i < size; ++i) {
        data[i - 1] = data[i];
      }

      data[--size] = 0;

      return value;
    }

    T get(int index) const {
      if(index < 0 || index >= size) throw std::runtime_error("Index out of bounds");

      return data[index];
    }

    int indexOf(T value) const {
      for(int i = 0; i < size; ++i) {
        if(data[i] == value)
          return i;
      }

      return -1;
    }

    bool contains(T value) const {
      return indexOf(value) != -1;
    }
};

template <typename T>
std::ostream& operator<<(std::ostream& os, const DArray<T>& arr) {
  os << "[";

  for(int i = 0; i < arr.getSize(); ++i) 
    os << arr.get(i) << " ";

  os << "]";

  return os;
}


int main(int argc, char** argv) {
  DArray<int> arr(2);

  arr.add(5);
  arr.add(3);
  arr.add(7);
  arr.add(512);
  arr.add(55);
  arr.add(124);

  std::cout << arr << std::endl;
  std::cout << arr.indexOf(512) << std::endl;

  arr.remove(arr.indexOf(512));

  std::cout << arr << std::endl;
  std::cout << arr.contains(124) << std::endl;
  std::cout << arr.getCapacity() << std::endl;
  std::cout << arr.getSize() << std::endl;
  std::cout << "------------" << std::endl;

  DArray<int> arr2 = arr;
  arr2.add(1024);
  std::cout << arr << std::endl;
  std::cout << arr2 << std::endl;
}